### PR TITLE
Prevent S3 URL presigning from generating a remote request span

### LINKED ISSUE #922: Redis, S3 with Unexpected Results
I have a code path that looks like this:

```ruby
def signed_url
  presigned_url = ->(_name) { Aws::S3::Object.new(...).presigned_url(:get, expires_in: ttl.to_i) }
  if FeatureSwitcher.enabled?(:cache_s3_signed_url)
    Rails.cache.fetch("signed_url_#{public_id}", expires_in: ttl, force: skip_cache, race_condition_ttl: 10.seconds, &presigned_url)
  else
    presigned_url.call(nil)
  end
end
```

When I turn on the feature switch, I see traces in DD flame graph UI that looks like this:

Screen Shot 2020-01-14 at 11 42 32 AM

When I turn off the feature switch, I no longer see any AWS commands like`s3.get_object` in the span list. 

I think the first screen shot is a bug, [`presigned_url`](https://github.com/aws/aws-sdk-ruby/blob/24fa03ecfb5a645e5ee0cd88cd7aae55c88247b9/gems/aws-sdk-s3/lib/aws-sdk-s3/customizations/object.rb#L185) makes no external calls to AWS and I'm unsure why using a cache is making it look like it is.

ddtrace 0.31.0
aws-sdk-s3 1.23.0
rails 5.2.2.1

### PR DESCRIPTION
**Pull request by marcotc**
Fixes #922

This PR addresses `ddtrace` capturing a "fake" S3 request that the AWS S3 SDK reports when [presigning a request URL](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Presigner.html#presigned_request-instance_method).

Presigned URLs give users a pre-authenticated URL to be used in the future to retrieve data from S3, but they do not preform any data retrieval to get created.

The Ruby S3 SDK performs a "fake" request internally to generated a presigned URL: it creates an S3 `get_object` request, sends it through the regular request handler (one of which being `Datadog::Contrib::Aws::Handler`), but hijacks the "data send" step at the end by not actually sending the request.

Unfortunately, there's no data we can use to differentiate between a real request and a fake request in our `ddtrace` handler. I've opened a PR with AWS to provide us with that information, so we can simplify the logic from this PR in the future: https://github.com/aws/aws-sdk-ruby/pull/2513

This PR effectively removes our request handler from these fake presign requests, the same approach taken internally by the S3 SDK for some of its own handlers that do not play nice with the fake presign request: https://github.com/aws/aws-sdk-ruby/blob/656691b470bb970627286cf1df2bd84049b6787d/gems/aws-sdk-s3/lib/aws-sdk-s3/presigner.rb#L194-L196

### CODE REVIEW

**Comment by allcentury** [comment]
@marcotc congrats on getting the AWS SDK updated as well, this looks great.  We're fine with this solution, or one where a AWS SDK minimum version is required.

File: lib/ddtrace/contrib/aws/instrumentation.rb [hunk 1]
```diff
@@ -56,6 +56,29 @@ def configuration
           Datadog.configuration[:aws]
         end
       end
+
+      # Removes API request instrumentation from S3 Presign URL creation.
+      #
+      # This is necessary because the the S3 SDK invokes the same handler
```
Comment (ericmustin) [inline lib/ddtrace/contrib/aws/instrumentation.rb:62]:
```suggestion
      # This is necessary because the S3 SDK invokes the same handler
```

**Review by ericmustin — DISMISSED** [review]
I think this is good to go, one small nit in code comments, not blocking though. 

It looks like your upstream PR got merged but aws doesn't cut a new release that frequently, however in the future we may want to revist this and rely on the upstream approach of checking `context[:presigned_url]` for relevant versions.

**Comment by marcotc** [comment]
> It looks like your upstream PR got merged but aws doesn't cut a new release that frequently, however in the future we may want to revist this and rely on the upstream approach of checking context[:presigned_url] for relevant versions.

It would be nice.
We still have to support clients with older versions though...

I'll add a comment referencing the easier way to do it in the code.

**Pull request merged**